controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -86,15 +113,88 @@
         this.counter = new AtomicInt96();
         this.limit = new AtomicInt96();
         this.refreshFutureRef = null;
+        this.completedTxnGC = new ZKGarbageCollector(COMPLETED_TXN_GC_NAME, storeHelper, this::gcCompletedTxn, gcPeriod);
+        this.completedTxnGC.startAsync();
+        this.completedTxnGC.awaitRunning();
     }
 
     private void initialize() {
         METRICS_PROVIDER.start();
     }
 
+    private CompletableFuture<Void> gcCompletedTxn() {
+        return storeHelper.getChildren(COMPLETED_TX_BATCH_ROOT_PATH)
+                .thenApply(children -> {
+                            // retain latest two and delete remainder.
+                            List<Long> list = children.stream().map(Long::parseLong).sorted().collect(Collectors.toList());
+                            if (list.size() > 2) {
```

Comment:
Why 2 and not 10? It seems like this number is heavily dependent on what the batching interval is. It also seems like we might want to configure the retention period on a per-stream basis. After all if a writer using a TXN will get an error and not be able to tell the txn status, it seems like there might be some user configurable way to say when that can occur. 
Ideally we would just piggyback on the existing stream retention period.